README.md
# VS Code Extension for Firebase

This extension allows you to send your project data to Firebase directly from Visual Studio Code.

## Features

- Send project data to Firebase
- Track project name, language, and status
- Automatic timer to update Firebase data
- Firebase configuration file management

## Requirements

Before using this extension, make sure you have the following prerequisites:

- Visual Studio Code version 1.79.0 or higher
- Firebase account and project

## Installation

1. Launch Visual Studio Code.
2. Go to the Extensions view by clicking on the square icon on the left sidebar or pressing `Ctrl+Shift+X`.
3. Search for "VS Code Extension for Firebase" and click **Install**.
4. After installation, click **YES** to `Do you want to create a new Firebase configuration file?` dialog box to configure the Firebase extension.

## Extension Settings

The extension contributes the following settings:

- `Send to Firebase: Open Config File`: Edit the firebase file.
- `Send to FIrebase: Reload App`: Edit the firebase file.

## Usage

1. Open a project folder in Visual Studio Code.
2. The extension will automatically track the project name and the programming language of the active file.
3. Firebase data will be sent to the configured Firebase database.
4. You can open the Firebase configuration file by running the **Send to Firebase: Open Config File** command from the Command Palette.

## Known Issues

- None at the moment. Please report any issues on the [GitHub repository](https://github.com/dinethsiriwardana/send-to-firebase).

## Release Notes

### 0.0.0

- Initial release of the VS Code Extension for Firebase.

### 0.0.1

- Added automatic timer to update Firebase data.
- Improved Firebase configuration file management.

### 0.0.2

- First release on the Visual Studio Marketplace.

### 0.0.3

- Fixed some bugs.

### 0.0.4

- Add 'Send to Firebase: Reload App' for reloading manually when configuration is complete.

### 1.0.0

- First stable release.

## Contributing

Contributions are welcome! Follow these steps to contribute to the project:

1. Fork the repository.
2. Create a new branch for your feature or bug fix.
3. Make the necessary changes and commit your code.
4. Submit a pull request explaining your changes.

For more information about Visual Studio Code's extension development, refer to the official [Extension API documentation](https://code.visualstudio.com/api).

**Enjoy using the VS Code Extension for Firebase!**


CHANGELOG.md
# Change Log

All notable changes to the "vs-code-ectension-for-firebase" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

out/test.js
"use strict";
// import * as vscode from "vscode";
// import { initializeApp } from "firebase/app";
// import { getDatabase, ref, set } from "firebase/database";
// const firebaseConfig = {
//   apiKey: "AIzaSyCSl3y9KF8w8zR5wXB0Nqu2sN_0ZKgb8nw",
//   authDomain: "vscodee.firebaseapp.com",
//   databaseURL: "https://vscodee-default-rtdb.firebaseio.com",
//   projectId: "vscodee",
//   storageBucket: "vscodee.appspot.com",
//   messagingSenderId: "991692892077",
//   appId: "1:991692892077:web:0a990adbd72b525a42e324",
// };
// const app = initializeApp(firebaseConfig);
// let timer: NodeJS.Timeout | null = null;
// const TIMEOUT_DURATION = 60 * 3 * 1000; // 3 minutes in milliseconds
// export function activate(context: vscode.ExtensionContext) {
//   console.log(
//     'Congratulations, your extension "vs-code-ectension-for-firebase" is now active!'
//   );
//   let disposable = vscode.commands.registerCommand(
//     "vs-code-ectension-for-firebase.helloWorld",
//     () => {
//       vscode.window.showInformationMessage(
//         "Hello World from VS code Extension for Firebase!"
//       );
//     }
//   );
//   context.subscriptions.push(disposable);
//   vscode.workspace.onDidOpenTextDocument(handleDocumentEventOpen);
//   vscode.workspace.onDidChangeTextDocument(handleDocumentEventChange);
// }
// function handleDocumentEventOpen(event: vscode.TextDocument) {
//   const db = getDatabase(app);
//   const editor = vscode.window.activeTextEditor;
//   if (editor) {
//     try {
//       const workspaceFolder = vscode.workspace.getWorkspaceFolder(event.uri);
//       const projectName = workspaceFolder ? workspaceFolder.name : "Unknown";
//       sendToFirebase(projectName, 1, "");
//       clearTimer();
//       startTimer(db);
//     } catch (error) {
//       vscode.window.showErrorMessage(
//         `Error handling document open event: ${error}`
//       );
//       console.error("Error handling document open event:", error);
//     }
//   }
// }
// function handleDocumentEventChange(event: vscode.TextDocumentChangeEvent) {
//   const db = getDatabase(app);
//   const editor = vscode.window.activeTextEditor;
//   if (editor) {
//     try {
//       const workspaceFolder = vscode.workspace.getWorkspaceFolder(
//         event.document.uri
//       );
//       const projectName = workspaceFolder ? workspaceFolder.name : "Unknown";
//       const language = event.document.languageId;
//       console.log(language);
//       sendToFirebase(projectName, 1, language);
//       clearTimer();
//       startTimer(db);
//     } catch (error) {
//       vscode.window.showErrorMessage(
//         `Error handling document change event: ${error}`
//       );
//     }
//   }
// }
// function sendToFirebase(projectName: string, stts: number, language: string) {
//   const db = getDatabase(app);
//   const currentTime = getCurrentTime();
//   try {
//     set(ref(db, "vscode_data/"), {
//       projectName: projectName,
//       language: language,
//       stts: stts,
//       time: currentTime,
//     });
//   } catch (error) {
//     vscode.window.showErrorMessage(`Error sending data to Firebase: ${error}`);
//   }
// }
// function startTimer(db: any) {
//   timer = setTimeout(() => {
//     console.log("Timer expired. Sending filename as 0");
//     try {
//       sendToFirebase("", 0, "");
//     } catch (error) {
//       vscode.window.showErrorMessage(
//         `Error sending data to Firebase: ${error}`
//       );
//     }
//   }, TIMEOUT_DURATION);
// }
// function clearTimer() {
//   if (timer) {
//     clearTimeout(timer);
//     timer = null;
//   }
// }
// function getCurrentTime(): string {
//   const currentDate = new Date();
//   const hours = String(currentDate.getHours()).padStart(2, "0");
//   const minutes = String(currentDate.getMinutes()).padStart(2, "0");
//   const seconds = String(currentDate.getSeconds()).padStart(2, "0");
//   return `${hours}:${minutes}:${seconds}`;
// }
// export function deactivate() {
//   clearTimer();
//   const db = getDatabase(app);
//   try {
//     sendToFirebase("", 0, "");
//   } catch (error) {
//     vscode.window.showErrorMessage(`Error sending data to Firebase: ${error}`);
//   }
// }
//# sourceMappingURL=test.js.map

src/extension.ts
import * as vscode from "vscode";
import { initializeApp } from "firebase/app";
import { getDatabase, ref, set } from "firebase/database";
import * as fs from "fs";
import * as path from "path";

let app: any;
let timer: NodeJS.Timeout | null = null;
const TIMEOUT_DURATION = 60 * 3 * 1000; // 3 minutes in milliseconds

// Global Firebase configuration

let firebaseConfig: any = null;

export function activate(context: vscode.ExtensionContext) {
  console.log(
    'Congratulations, your extension "vs-code-ectension-for-firebase" is now active!'
  );

  let openConfigFileDisposable = vscode.commands.registerCommand(
    "send-to-firebase.openConfigFile",
    () => {
      const configPath = path.join(
        context.extensionPath,
        "firebase-config.json"
      );
      vscode.workspace.openTextDocument(configPath).then((document) => {
        vscode.window.showTextDocument(document);
      });
    }
  );

  let reloadAppDisposable = vscode.commands.registerCommand(
    "send-to-firebase.reloadApp",
    () => {
      console.log("Reloading the app...");
      vscode.window.showWarningMessage("Reloading the app...");

      // Perform any necessary cleanup or reset operations here

      // Load Firebase config if the config file exists
      if (fs.existsSync(configPath)) {
        loadFirebaseConfig(configPath);
      } else {
        vscode.window.showWarningMessage(
          "Firebase configuration file not found"
        );
      }
    }
  );

  context.subscriptions.push(openConfigFileDisposable, reloadAppDisposable);
  vscode.workspace.onDidOpenTextDocument(handleDocumentEventOpen);
  vscode.workspace.onDidChangeTextDocument(handleDocumentEventChange);

  const configPath = path.join(context.extensionPath, "firebase-config.json");
  checkFirebaseConfigFile(configPath);

  const watcher = vscode.workspace.createFileSystemWatcher(configPath);

  watcher.onDidChange(() => {
    console.log("Firebase config file changed. Reloading the app...");
    vscode.window.showWarningMessage(
      "Firebase config file changed. Reloading the app..."
    );
    loadFirebaseConfig(configPath);
    // You can call any necessary functions or perform actions after the config file is changed.
  });

  // Load Firebase config if the config file exists
  if (fs.existsSync(configPath)) {
    loadFirebaseConfig(configPath);
  } else {
    vscode.window.showWarningMessage("Firebase configuration file not found");
  }
}

function checkFirebaseConfigFile(configPath: string) {
  if (firebaseConfig) {
    app = initializeApp(firebaseConfig);
  } else if (!fs.existsSync(configPath)) {
    vscode.window
      .showInformationMessage(
        "Do you want to create a new Firebase configuration file?",
        "Yes",
        "No"
      )
      .then((choice) => {
        if (choice === "Yes") {
          const configTemplate = {
            apiKey: "YOUR_API_KEY",
            authDomain: "YOUR_AUTH_DOMAIN",
            databaseURL: "YOUR_DATABASE_URL",
            projectId: "YOUR_PROJECT_ID",
            storageBucket: "YOUR_STORAGE_BUCKET",
            messagingSenderId: "YOUR_MESSAGING_SENDER_ID",
            appId: "YOUR_APP_ID",
          };

          fs.writeFileSync(configPath, JSON.stringify(configTemplate, null, 2));

          vscode.workspace.openTextDocument(configPath).then((document) => {
            vscode.window.showTextDocument(document);
          });
          vscode.window.showWarningMessage(
            "Run `Send to Firebase: Reload App` to Reload after configuring Firebase (F1 -> Send to Firebase: Reload App)"
          );
        }
      });
  }
}

function loadFirebaseConfig(configPath: string) {
  const configContent = fs.readFileSync(configPath, "utf8");
  try {
    const config = JSON.parse(configContent);
    app = initializeApp(config);
    firebaseConfig = config;
    console.log("Firebase app initialized with the updated configuration");
  } catch (error) {
    vscode.window.showErrorMessage("Invalid Firebase configuration");
  }
}

function handleDocumentEventOpen(event: vscode.TextDocument) {
  const db = getDatabase(app);
  const editor = vscode.window.activeTextEditor;
  if (editor) {
    try {
      const workspaceFolder = vscode.workspace.getWorkspaceFolder(event.uri);
      const projectName = workspaceFolder ? workspaceFolder.name : "Unknown";
      sendToFirebase(projectName, 1, "");

      clearTimer();
      startTimer(db);
    } catch (error) {
      vscode.window.showErrorMessage(
        `Error handling document open event: ${error}`
      );
      console.error("Error handling document open event:", error);
    }
  }
}

function handleDocumentEventChange(event: vscode.TextDocumentChangeEvent) {
  const db = getDatabase(app);
  const editor = vscode.window.activeTextEditor;
  if (editor) {
    try {
      const workspaceFolder = vscode.workspace.getWorkspaceFolder(
        event.document.uri
      );
      const projectName = workspaceFolder ? workspaceFolder.name : "Unknown";

      const language = event.document.languageId;

      console.log(language);

      sendToFirebase(projectName, 1, language);

      clearTimer();
      startTimer(db);
    } catch (error) {
      vscode.window.showErrorMessage(
        `Error handling document change event: ${error}`
      );
      console.error("Error handling document change event:", error);
    }
  }
}

function sendToFirebase(projectName: string, stts: number, language: string) {
  const db = getDatabase(app);
  const currentTime = getCurrentTime();
  try {
    set(ref(db, "vscode_data/"), {
      projectName: projectName,
      language: language,
      stts: stts,
      time: currentTime,
    });
  } catch (error) {
    vscode.window.showErrorMessage(`Error sending data to Firebase: ${error}`);
    console.error("Error sending data to Firebase:", error);
  }
}

function startTimer(db: any) {
  timer = setTimeout(() => {
    console.log("Timer expired. Sending filename as 0");
    try {
      sendToFirebase("", 0, "");
    } catch (error) {
      vscode.window.showErrorMessage(
        `Error sending data to Firebase: ${error}`
      );
      console.error("Error sending data to Firebase:", error);
    }
  }, TIMEOUT_DURATION);
}

function clearTimer() {
  if (timer) {
    clearTimeout(timer);
    timer = null;
  }
}

function getCurrentTime(): string {
  const currentDate = new Date();
  const hours = String(currentDate.getHours()).padStart(2, "0");
  const minutes = String(currentDate.getMinutes()).padStart(2, "0");
  const seconds = String(currentDate.getSeconds()).padStart(2, "0");
  return `${hours}:${minutes}:${seconds}`;
}

export function deactivate() {
  clearTimer();
  const db = getDatabase(app);

  try {
    sendToFirebase("", 0, "");
  } catch (error) {
    vscode.window.showErrorMessage(`Error sending data to Firebase: ${error}`);
    console.error("Error sending data to Firebase:", error);
  }
}


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const app_1 = require("firebase/app");
const database_1 = require("firebase/database");
const fs = require("fs");
const path = require("path");
let app;
let timer = null;
const TIMEOUT_DURATION = 60 * 3 * 1000; // 3 minutes in milliseconds
// Global Firebase configuration
let firebaseConfig = null;
function activate(context) {
    console.log('Congratulations, your extension "vs-code-ectension-for-firebase" is now active!');
    let openConfigFileDisposable = vscode.commands.registerCommand("send-to-firebase.openConfigFile", () => {
        const configPath = path.join(context.extensionPath, "firebase-config.json");
        vscode.workspace.openTextDocument(configPath).then((document) => {
            vscode.window.showTextDocument(document);
        });
    });
    let reloadAppDisposable = vscode.commands.registerCommand("send-to-firebase.reloadApp", () => {
        console.log("Reloading the app...");
        vscode.window.showWarningMessage("Reloading the app...");
        // Perform any necessary cleanup or reset operations here
        // Load Firebase config if the config file exists
        if (fs.existsSync(configPath)) {
            loadFirebaseConfig(configPath);
        }
        else {
            vscode.window.showWarningMessage("Firebase configuration file not found");
        }
    });
    context.subscriptions.push(openConfigFileDisposable, reloadAppDisposable);
    vscode.workspace.onDidOpenTextDocument(handleDocumentEventOpen);
    vscode.workspace.onDidChangeTextDocument(handleDocumentEventChange);
    const configPath = path.join(context.extensionPath, "firebase-config.json");
    checkFirebaseConfigFile(configPath);
    const watcher = vscode.workspace.createFileSystemWatcher(configPath);
    watcher.onDidChange(() => {
        console.log("Firebase config file changed. Reloading the app...");
        vscode.window.showWarningMessage("Firebase config file changed. Reloading the app...");
        loadFirebaseConfig(configPath);
        // You can call any necessary functions or perform actions after the config file is changed.
    });
    // Load Firebase config if the config file exists
    if (fs.existsSync(configPath)) {
        loadFirebaseConfig(configPath);
    }
    else {
        vscode.window.showWarningMessage("Firebase configuration file not found");
    }
}
exports.activate = activate;
function checkFirebaseConfigFile(configPath) {
    if (firebaseConfig) {
        app = (0, app_1.initializeApp)(firebaseConfig);
    }
    else if (!fs.existsSync(configPath)) {
        vscode.window
            .showInformationMessage("Do you want to create a new Firebase configuration file?", "Yes", "No")
            .then((choice) => {
            if (choice === "Yes") {
                const configTemplate = {
                    apiKey: "YOUR_API_KEY",
                    authDomain: "YOUR_AUTH_DOMAIN",
                    databaseURL: "YOUR_DATABASE_URL",
                    projectId: "YOUR_PROJECT_ID",
                    storageBucket: "YOUR_STORAGE_BUCKET",
                    messagingSenderId: "YOUR_MESSAGING_SENDER_ID",
                    appId: "YOUR_APP_ID",
                };
                fs.writeFileSync(configPath, JSON.stringify(configTemplate, null, 2));
                vscode.workspace.openTextDocument(configPath).then((document) => {
                    vscode.window.showTextDocument(document);
                });
                vscode.window.showWarningMessage("Run `Send to Firebase: Reload App` to Reload after configuring Firebase (F1 -> Send to Firebase: Reload App)");
            }
        });
    }
}
function loadFirebaseConfig(configPath) {
    const configContent = fs.readFileSync(configPath, "utf8");
    try {
        const config = JSON.parse(configContent);
        app = (0, app_1.initializeApp)(config);
        firebaseConfig = config;
        console.log("Firebase app initialized with the updated configuration");
    }
    catch (error) {
        vscode.window.showErrorMessage("Invalid Firebase configuration");
    }
}
function handleDocumentEventOpen(event) {
    const db = (0, database_1.getDatabase)(app);
    const editor = vscode.window.activeTextEditor;
    if (editor) {
        try {
            const workspaceFolder = vscode.workspace.getWorkspaceFolder(event.uri);
            const projectName = workspaceFolder ? workspaceFolder.name : "Unknown";
            sendToFirebase(projectName, 1, "");
            clearTimer();
            startTimer(db);
        }
        catch (error) {
            vscode.window.showErrorMessage(`Error handling document open event: ${error}`);
            console.error("Error handling document open event:", error);
        }
    }
}
function handleDocumentEventChange(event) {
    const db = (0, database_1.getDatabase)(app);
    const editor = vscode.window.activeTextEditor;
    if (editor) {
        try {
            const workspaceFolder = vscode.workspace.getWorkspaceFolder(event.document.uri);
            const projectName = workspaceFolder ? workspaceFolder.name : "Unknown";
            const language = event.document.languageId;
            console.log(language);
            sendToFirebase(projectName, 1, language);
            clearTimer();
            startTimer(db);
        }
        catch (error) {
            vscode.window.showErrorMessage(`Error handling document change event: ${error}`);
            console.error("Error handling document change event:", error);
        }
    }
}
function sendToFirebase(projectName, stts, language) {
    const db = (0, database_1.getDatabase)(app);
    const currentTime = getCurrentTime();
    try {
        (0, database_1.set)((0, database_1.ref)(db, "vscode_data/"), {
            projectName: projectName,
            language: language,
            stts: stts,
            time: currentTime,
        });
    }
    catch (error) {
        vscode.window.showErrorMessage(`Error sending data to Firebase: ${error}`);
        console.error("Error sending data to Firebase:", error);
    }
}
function startTimer(db) {
    timer = setTimeout(() => {
        console.log("Timer expired. Sending filename as 0");
        try {
            sendToFirebase("", 0, "");
        }
        catch (error) {
            vscode.window.showErrorMessage(`Error sending data to Firebase: ${error}`);
            console.error("Error sending data to Firebase:", error);
        }
    }, TIMEOUT_DURATION);
}
function clearTimer() {
    if (timer) {
        clearTimeout(timer);
        timer = null;
    }
}
function getCurrentTime() {
    const currentDate = new Date();
    const hours = String(currentDate.getHours()).padStart(2, "0");
    const minutes = String(currentDate.getMinutes()).padStart(2, "0");
    const seconds = String(currentDate.getSeconds()).padStart(2, "0");
    return `${hours}:${minutes}:${seconds}`;
}
function deactivate() {
    clearTimer();
    const db = (0, database_1.getDatabase)(app);
    try {
        sendToFirebase("", 0, "");
    }
    catch (error) {
        vscode.window.showErrorMessage(`Error sending data to Firebase: ${error}`);
        console.error("Error sending data to Firebase:", error);
    }
}
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map